base/inc/AdePT/SparseVector.h

```diff
@@ -0,0 +1,746 @@
+      int idest = dest_offset + group_offset + subgroup_index + __popc(mask & ((1 << lnid) - 1));
+      int isrc  = 1024 * warp_id + 32 * group_id + lnid;
+      // printf("== copy from %d to %d\n", isrc, idest);
+      new (output->data() + idest) Type((*svect)[isrc]); // call in-place copy constructor starting with last element
+    }
+  }
+}
+
+/// Dynamic launcher for select_kernel
+template <typename Type, typename Predicate>
+__global__ void select_and_move_kernel_launcher(SparseVectorInterface<Type> *svect, Predicate pred_func,
+                                                SparseVectorInterface<Type> *output, unsigned *nselected)
+{
+  using Vector_t            = SparseVectorInterface<Type>;
+  const int num_items       = svect->size();
+  constexpr int num_threads = 128;
+  const int num_blocks      = (num_items + 4095) / 4096;
+  *nselected                = 0;
+  sa_detail::select_and_move_kernel<Type>
+      <<<num_blocks, num_threads>>>(svect, pred_func, output, nselected, output->size());
+}
+
+#endif // COPCORE_CUDA_COMPILER
+} // end namespace sa_detail
+
+/** @brief SparseVector interface */
+template <typename Type>
+class SparseVectorInterface {
+  using Vector_t               = SparseVectorInterface<Type>;
+  using value_type             = Type;
+  using pointer                = value_type *;
+  using const_pointer          = const value_type *;
+  using reference              = value_type &;
+  using const_reference        = const value_type &;
+  using iterator               = value_type *;
+  using const_iterator         = const value_type *;
+  using reverse_iterator       = std::reverse_iterator<iterator>;
+  using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+  using size_t                 = std::size_t;
+  using Index_t                = unsigned int;
+  using Mask_t                 = unsigned int;
+  using AtomicInt_t            = adept::Atomic_t<int>;
+  using AtomicMask_t           = adept::Atomic_t<Mask_t>;
+  using LaunchGrid_t           = copcore::launch_grid<copcore::BackendType::CUDA>;
+
+protected:
+  void *fBegin     = nullptr; ///< Start address of the vector data
+  void *fMaskBegin = nullptr; ///< Start address of the mask data
+  AtomicInt_t fNbooked;       ///< Number of booked elements (requested concurrently, may become larger than N)
+  AtomicInt_t fNshared;       ///< Number of shared elements (given away, decreases only after clear/compact)
+  AtomicInt_t fNused;         ///< Number of elements currently in use
+  unsigned fCapacity{0};      ///< Capacity of the vector
+  unsigned fMaskCapacity{0};  ///< Capacity of the mask vector
+  int fNumSMs{0};             ///< number of streaming multi-processors
+
+  /// Set first element address
+  __host__ __device__ __forceinline__ void setFirstEl(void *addr) { fBegin = addr; }
+
+  /// Set first mask address
+  __host__ __device__ __forceinline__ void setFirstElMask(void *addr) { fMaskBegin = addr; }
```

Comment:
Not needed, the derived class has direct access to the pointers

Reply:
leftovers - removed